design/general-progress-monitoring.md

```diff
@@ -0,0 +1,462 @@
+"WaitingForPluginOperations" or "WaitingForPluginOperationsPartialFailure" state.  A backup/restore
+which will be marked "Failed" will go directly to the "Failed" phase.  Uploads may continue in the
+background for snapshots that were taken by a "Failed" backup/restore, but no progress will not be
+monitored or updated.  When a "Failed" backup is deleted, all snapshots will be deleted and at that
+point any uploads still in progress should be aborted.
+
+### WaitingForPluginOperations (new)
+The "WaitingForPluginOperations" phase signifies that the main part of the backup/restore, including
+snapshotting has completed successfully and uploading and any other asynchronous BIA/RIA plugin
+operations are continuing.  In the event of an error during this phase, the phase will change to
+WaitingForPluginOperationsPartialFailure.  On success, the phase changes to Completed.  Backups
+cannot be restored from when they are in the WaitingForPluginOperations state.
+
+### WaitingForPluginOperationsPartialFailure (new)
+The "WaitingForPluginOperationsPartialFailure" phase signifies that the main part of the
+backup/restore, including snapshotting has completed, but there were partial failures either during
+the main part or during any async operations, including snapshot uploads.  Backups cannot be
+restored from when they are in the WaitingForPluginOperationsPartialFailure state.
+
+### Failed
+When a backup/restore has had fatal errors it is marked as "Failed" Backups in this state cannot be
+restored from.
+
+### Completed
+The "Completed" phase signifies that the backup/restore has completed, all data has been transferred
+to stable storage (or restored to the cluster) and any backup in this state is ready to be used in a
+restore.  When the Completed phase has been reached for a backup it is safe to remove any of the
+items that were backed up.
+
+### PartiallyFailed
+The "PartiallyFailed" phase signifies that the backup/restore has completed and at least part of the
+backup/restore is usable.  Restoration from a PartiallyFailed backup will not result in a complete
+restoration but pieces may be available.
+
+## Workflow
+
+When a Backup or Restore Action is executed, any BackupItemAction, RestoreItemAction, or
+VolumeSnapshot plugins will return operation IDs (snapshot IDs or other plugin-specific
+identifiers).  The plugin should be able to provide status on the progress for the snapshot and
+handle cancellation of the operation/upload if the snapshot is deleted.  If the plugin is restarted,
+the operation ID should remain valid.
+
+When all snapshots have been taken and Kubernetes resources have been persisted to the ObjectStorePlugin
+the backup will either have fatal errors or will be at least partially usable.
+
+If the backup/restore has fatal errors it will move to the "Failed" state and finish. If a
+backup/restore fails, the upload or other operation will not be cancelled but it will not be
```

Comment:
As previous comment, my preference is not starting upload in the case, not sure about other users or maintainers' opinion.

Reply:
so what can be done to mitigate the side-effects for the async operations that have been started when the backup goes to `failed` state?

Reply:
@reasonerjt So currently, once the backup has started, there's not really anything that can happen to move the backup to Failed state (vs. PartiallyFailed). If a running backup (with running async actions) goes into partially failed state, that doesn't really have any impact on async operations though. If, for example, a datamover plugin is copying a completed VolumeSnapshotContents to the BSL, a failure in the associated backup isn't all that relevant. If there's a partial failure, Velero will continue to monitor and move it from WaitingForPluginOperationsPartiallyFailed to PartiallyFailed once done. If a backup did end up moving to an actual complete Failed case (which, again, isn't possible with current workflow), then the async operation would continue in the background, but velero just wouldn't check back on it again. So I don't think there's anything to mitigate here. In any case, I don't think that's particularly relevant to my point above, which is that plugins should start their async operations immediately when running Execute, since the whole point of this is to start the operation now and not block ongoing backup actions while this long-running plugin action is taking place.